src/SparkTestBase.sol

```diff
@@ -40,22 +45,133 @@ interface IExecutable {
     function execute() external;
 }
 
-abstract contract SparkTestBase is Test {
-    address internal executor;
-    address internal payload;
+abstract contract SpellRunner is Test {
+    using DomainHelpers for Domain;
+    using DomainHelpers for StdChains.Chain;
+
+    struct ChainSpellMetadata{
+      address   payload;
+      IExecutor executor;
+      Domain    domain;
+      /// @notice on mainnet: zero
+      /// on L2s: mainnet address of the bridge that'll include txs in the L2
+      Bridge    bridge;
+      // @notice coupled to SparklendTests, zero on chains where sparklend is not present
+      IPoolAddressesProviderRegistry sparklendPooAddressProviderRegistry;
+    }
+
+    mapping(ChainId chainId => ChainSpellMetadata chainSpellMetadata) internal chainSpellMetadata;
+
+    ChainId[] internal allChains;
+    string internal    id;
+
+    modifier onChain(ChainId chainId) virtual {
+        ChainId currentChain = ChainIdUtils.fromUint(block.chainid);
+        chainSpellMetadata[chainId].domain.selectFork();
+        _;
+        chainSpellMetadata[currentChain].domain.selectFork();
+    }
+
+    /// @dev to be called in setUp
+    function setupDomains(uint256 mainnetForkBlock, uint256 baseForkBlock, uint256 gnosisForkBlock) internal {
+        chainSpellMetadata[ChainIdUtils.Ethereum()].domain = getChain("mainnet").createFork(mainnetForkBlock);
+        chainSpellMetadata[ChainIdUtils.Base()].domain     = getChain("base").createFork(baseForkBlock);
+        chainSpellMetadata[ChainIdUtils.Gnosis()].domain   = getChain("gnosis_chain").createFork(gnosisForkBlock);
+
+        chainSpellMetadata[ChainIdUtils.Ethereum()].executor = IExecutor(Ethereum.SPARK_PROXY);
+        chainSpellMetadata[ChainIdUtils.Base()].executor     = IExecutor(Base.SPARK_EXECUTOR);
+        chainSpellMetadata[ChainIdUtils.Gnosis()].executor   = IExecutor(Gnosis.AMB_EXECUTOR);
+
+        chainSpellMetadata[ChainIdUtils.Base()].bridge = OptimismBridgeTesting.createNativeBridge(chainSpellMetadata[ChainIdUtils.Ethereum()].domain, chainSpellMetadata[ChainIdUtils.Base()].domain);
+
+        chainSpellMetadata[ChainIdUtils.Ethereum()].sparklendPooAddressProviderRegistry = IPoolAddressesProviderRegistry(Ethereum.POOL_ADDRESSES_PROVIDER_REGISTRY);
+        chainSpellMetadata[ChainIdUtils.Gnosis()].sparklendPooAddressProviderRegistry   = IPoolAddressesProviderRegistry(Gnosis.POOL_ADDRESSES_PROVIDER_REGISTRY);
+
+        allChains.push(ChainIdUtils.Ethereum());
+        allChains.push(ChainIdUtils.Base());
+        allChains.push(ChainIdUtils.Gnosis());
+    }
+
+    function spellIdentifier(ChainId chainId) private view returns(string memory){
+        string memory slug            = string(abi.encodePacked("Spark", chainId.toDomainString(), "_", id));
+        string memory spellIdentifier = string(abi.encodePacked(slug, ".sol:", slug));
```

Comment:
This is shadowing the function name. Add an underscore or rename.

Reply:
fixed in 81de134